File: cowork-roadmap/src/main/java/com/cowork/roadmap/global/team/ProjectionTopicIdentityProvider.java
package com.cowork.roadmap.global.team;

import java.time.Duration;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import org.apache.kafka.clients.admin.Admin;
import org.apache.kafka.clients.admin.OffsetSpec;
import org.apache.kafka.common.TopicPartition;
import org.apache.kafka.common.Uuid;
import org.springframework.boot.kafka.autoconfigure.KafkaProperties;
import org.springframework.stereotype.Component;

import jakarta.annotation.PreDestroy;

@Component
public class ProjectionTopicIdentityProvider {

    private static final long METADATA_TIMEOUT_SECONDS = 10L;

    private final Admin admin;

    public ProjectionTopicIdentityProvider(KafkaProperties kafkaProperties) {
        this.admin = Admin.create(kafkaProperties.buildAdminProperties());
    }

    String topicId(String topic) {
        try {
            var description = admin.describeTopics(java.util.List.of(topic))
                    .allTopicNames()
                    .get(METADATA_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                    .get(topic);
            if (description == null || description.topicId().equals(Uuid.ZERO_UUID)) {
                throw new IllegalStateException("Kafka topic ID is unavailable: " + topic);
            }
            return description.topicId().toString();
        } catch (Exception exception) {
            if (exception instanceof InterruptedException) {
                Thread.currentThread().interrupt();
            }
            throw new IllegalStateException("Failed to load Kafka topic ID: " + topic, exception);
        }
    }

    ProjectionTopicState topicState(String topic) {
        try {
            var description = admin.describeTopics(java.util.List.of(topic))
                    .allTopicNames()
                    .get(METADATA_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                    .get(topic);
            if (description == null || description.topicId().equals(Uuid.ZERO_UUID)) {
                throw new IllegalStateException("Kafka topic ID is unavailable: " + topic);
            }
            Collection<TopicPartition> partitions = description.partitions()
                    .stream()
                    .map(info -> new TopicPartition(topic, info.partition()))
                    .toList();
            Map<TopicPartition, OffsetSpec> earliestRequest = new HashMap<>();
            Map<TopicPartition, OffsetSpec> latestRequest = new HashMap<>();
            partitions.forEach(partition -> {
                earliestRequest.put(partition, OffsetSpec.earliest());
                latestRequest.put(partition, OffsetSpec.latest());
            });
            var beginnings = admin.listOffsets(earliestRequest).all().get(METADATA_TIMEOUT_SECONDS, TimeUnit.SECONDS);
            var ends = admin.listOffsets(latestRequest).all().get(METADATA_TIMEOUT_SECONDS, TimeUnit.SECONDS);
            Map<Integer, ProjectionBrokerRange> ranges = new HashMap<>();
            partitions.forEach(partition -> ranges.put(partition.partition(),
                    new ProjectionBrokerRange(beginnings.get(partition).offset(), ends.get(partition).offset())));
            return new ProjectionTopicState(description.topicId().toString(), Map.copyOf(ranges));
        } catch (Exception exception) {
            if (exception instanceof InterruptedException) {
                Thread.currentThread().interrupt();
            }
            throw new IllegalStateException("Failed to load Kafka topic state: " + topic, exception);
        }
    }

    @PreDestroy
    public void close() {
        admin.close(Duration.ofSeconds(5));
    }
}

File: cowork-channel/src/main/kotlin/com/cowork/channel/global/projection/ProjectionTopicIdentityProvider.kt
package com.cowork.channel.global.projection

import jakarta.annotation.PreDestroy
import org.apache.kafka.clients.admin.Admin
import org.apache.kafka.clients.admin.OffsetSpec
import org.apache.kafka.common.TopicPartition
import org.apache.kafka.common.Uuid
import org.springframework.boot.kafka.autoconfigure.KafkaProperties
import org.springframework.stereotype.Component
import java.time.Duration
import java.util.concurrent.TimeUnit

@Component
class ProjectionTopicIdentityProvider(kafkaProperties: KafkaProperties) {
    private val admin = Admin.create(kafkaProperties.buildAdminProperties())

    fun topicId(topic: String): String {
        val description = requireNotNull(
            admin.describeTopics(listOf(topic)).allTopicNames().get(10, TimeUnit.SECONDS)[topic],
        ) { "Kafka topic metadata가 없습니다: $topic" }
        check(description.topicId() != Uuid.ZERO_UUID) { "Kafka topic ID를 확인할 수 없습니다: $topic" }
        return description.topicId().toString()
    }

    fun topicStates(topics: Set<String>): Map<String, ProjectionTopicState> {
        if (topics.isEmpty()) return emptyMap()
        val descriptions = admin.describeTopics(topics).allTopicNames().get(10, TimeUnit.SECONDS)
        val partitions = descriptions.flatMap { (topic, description) ->
            check(description.topicId() != Uuid.ZERO_UUID) { "Kafka topic ID를 확인할 수 없습니다: $topic" }
            description.partitions().map { TopicPartition(topic, it.partition()) }
        }
        val beginningOffsets = admin.listOffsets(partitions.associateWith { OffsetSpec.earliest() })
            .all().get(10, TimeUnit.SECONDS)
        val endOffsets = admin.listOffsets(partitions.associateWith { OffsetSpec.latest() })
            .all().get(10, TimeUnit.SECONDS)

        return descriptions.mapValues { (topic, description) ->
            ProjectionTopicState(
                topicId = description.topicId().toString(),
                ranges = description.partitions().associate { partitionInfo ->
                    val partition = TopicPartition(topic, partitionInfo.partition())
                    partition.partition() to ProjectionBrokerRange(
                        beginningOffset = requireNotNull(beginningOffsets[partition]).offset(),
                        endOffset = requireNotNull(endOffsets[partition]).offset(),
                    )
                },
            )
        }
    }

    @PreDestroy
    fun close() {
        admin.close(Duration.ofSeconds(5))
    }
}
